tempesta_fw/addr.c

```diff
@@ -205,87 +221,24 @@ tfw_inet_pton_ipv6(char **p, struct sockaddr_in6 *addr)
 /**
  * Parse IPv4 and IPv6 addresses with optional port.
  * See RFC5952.
- *
- * @p - string pointer, updated by the function.
- * @addr - distination to write as a pointer to a union of sockaddr_in and
- * 	   sockaddr_in6.
  */
 int
-tfw_inet_pton(char **p, void *addr)
+tfw_inet_pton(const char *str, TfwAddr *addr)
 {
-	int mode = 0;
-
-	/* Eat empty string prefix. */
-	while (**p && isspace(**p))
-		++*p;
+	memset(addr, 0, sizeof(*addr));
 
-	/* Determine type of the address (IPv4/IPv6). */
-	if (**p == '[' || isalpha(**p)) {
-		mode = 6;
-	} else {
-		char *p1 = *p;
-		while (*p1 && isdigit(*p1))
-			p1++;
-		if (*p1 == ':') {
-			mode = 6;
-		}
-		else if (*p1 == '.') {
-			mode = 4;
-		}
-		else {
-			TFW_ERR("bad string: %s\n", *p);
-			return -EINVAL;
-		}
-	}
+	/* The IPv6 address must be enclosed into square brackets,
+	 * or else we can't distinguish it from the port. */
```

Comment:
Read rfc5956 chapter 6 "Notes on Combining IPv6 Addresses with Port Numbers". The same as above: throw out the new and wrong code and adjust the previous right code. The previous code IIRC does not handle all the cases specified by the RFC, but it did the things better.